src/scheduler.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "scheduler.h"
+
+#include <boost/bind.hpp>
+#include <utility>
+
+CScheduler::CScheduler()
+{
+}
+
+
+boost::thread*
+CScheduler::run()
+{
+    boost::thread* t = new boost::thread(boost::bind(&CScheduler::SchedulerLoop, this));
+    return t;
+}
+
+void
+CScheduler::SchedulerLoop()
+{
+    boost::unique_lock<boost::mutex> lock(newTaskMutex);
+
+    // newTaskMutex is locked throughout this loop EXCEPT
+    // when the thread is waiting or when the user's function
+    // is called.
+    while (1)
+    {
+        if (taskQueue.empty())
+        {
+            // Wait until there is something to do.
+            newTaskScheduled.wait(lock);
+        }
+        else
+        {
+            // Wait until either there is a new task, or until
+            // the time of the first item on the queue:
+            while (!taskQueue.empty() &&
+                   newTaskScheduled.wait_until(lock, taskQueue.begin()->first) != boost::cv_status::timeout)
```

Comment:
Does this work when the timestamp of the first item in the queue is in the past?